### DISCUSSION TITLE
analogRead inside Timer ISR causes MBED OS to crash on RP2040 Connect

### DISCUSSION THREAD

**Discussion by jmdodd95682** [khoih-prog/MBED_RPI_PICO_TimerInterrupt#6] [General]
I been trying to get a HW timer running to periodically sample an analog input value.  I can get the timer working fine using the example code.  I set the timer to 30KHz timeout and inside the TimeoutHandler the only thing I have is:

input_signal=analogRead(INPUT_PIN);

Where INPUT_PIN is defined as A3.  I also tried A6 to see if it was related to WiFiNINA.h.

The first time it hits the TimeoutHandler, the MBED OS crashes and the RP2040 drops into "SOS" mode.

I tried delaying the analogRead thinking it might have something to do with setup.  But, it seems like (I have not verified) that the analogRead is using the same timers as the MBED_RPI_PICO_TimerInterrupt.  This is somehow causing the MBED OS to go off into the weeds. 

Is this a known problem?  Is this something that can be worked around.  Without the ability to do the analogRead, there is no point is trying to us the HW timers.

Any help would be appreciated.

**Comment by khoih-prog** [comment]
Hi @jmdodd95682 

> The first time it hits the TimeoutHandler, the MBED OS crashes and the RP2040 drops into "SOS" mode.

It's bad that you design to use the TimerInterrupt the wrong way. You violate the basic rule of using any interrupt.
Have a look at the `analogRead()` code of WiFiNINA

https://github.com/arduino-libraries/WiFiNINA/blob/0e7e65b6273007c70176c665b3ede857b28761e3/src/utility/wifi_drv.cpp#L1122-L1148

```
uint16_t WiFiDrv::analogRead(uint8_t adc_channel)
{
    WAIT_FOR_SLAVE_SELECT();
    // Send Command
    SpiDrv::sendCmd(GET_ANALOG_READ, PARAM_NUMS_1);
    SpiDrv::sendParam((uint8_t*)&adc_channel, 1, LAST_PARAM);


    // pad to multiple of 4
    SpiDrv::readChar();
    SpiDrv::readChar();


    SpiDrv::spiSlaveDeselect();
    //Wait the reply elaboration
    SpiDrv::waitForSlaveReady();
    SpiDrv::spiSlaveSelect();


    // Wait for reply
    uint16_t adc_raw = 0;
    uint8_t adc_raw_len = 0;
    if (!SpiDrv::waitResponseCmd(GET_ANALOG_READ, PARAM_NUMS_1, (uint8_t*)&adc_raw, &adc_raw_len))
    {
        WARN("error waitResponse");
    }
    SpiDrv::spiSlaveDeselect();

    return adc_raw;
}
```


The better way to do is to set a flag inside ISR, then poll, using `millis()` or even `micros()`. When the flag is set, call `analogRead()` and clear the flag.

Good Luck,

**Comment by jmdodd95682** [comment]
Thanks for the response!  I'm used to being able to perform atomic operations within the ISR to avoid asynchronous behavior.  However, I tried your suggestion and it does work.  I will have to watch my delay in the main loop carefully so that I am not losing samples, but I think it should work.  By the way, the timer works extremely well.  Thanks again!

**Comment by Garrabrant** [comment]
I discovered the same issue, but was able to resolve it by using adc_read(); instead of analogRead(INPUT_PIN);

I also needed the following before starting the interrupt:
    adc_init();
    adc_gpio_init(26);
    adc_select_input(0);

**Comment by jmdodd95682** [comment]
Interesting.  I am not familiar with adc_read().  I’ll try it too.  I like learning new things.

Thanks.

**Comment by jmdodd95682** [comment]
Thanks again for the help.  Its working perfectly.  BTW the adc_read is about 4 times faster than using the standard analogRead routine.  It only takes 2us versus 8us.

And, as you pointed out, it works inside the Timer ISR, so the implementation is a bit cleaner.
